File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
using namespace std;

int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(0);
    int n;
    stack <char> stack;
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        string s;
        cin >> s;
        bool check = true;
        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == '(')
                stack.push('(');
            else if (s[i] == ')')
            {
                if (!stack.empty())
                    stack.pop();
                else if(stack.empty())
                {
                    check = false;
                    break;
                }
            }
        }
        if (!stack.empty())
            check = false;
        if (check)
            cout << "YES"<<"\n";
        else
            cout << "NO"<<"\n";
        while (!stack.empty())
            stack.pop();
    }
    return 0;
}

File: 백준/Silver/9012. 괄호/괄호.java
/* ************************************************************************** */
/*                                                                            */
/*                                                      :::    :::    :::     */
/*   Problem Number: 9012                              :+:    :+:      :+:    */
/*                                                    +:+    +:+        +:+   */
/*   By: eogud3332 <boj.kr/u/eogud3332>              +#+    +#+          +#+  */
/*                                                  +#+      +#+        +#+   */
/*   https://boj.kr/9012                           #+#        #+#      #+#    */
/*   Solved: 2026/02/06 17:29:50 by eogud3332     ###          ###   ##.kr    */
/*                                                                            */
/* ************************************************************************** */

import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		while (N-- > 0)
			System.out.println(check(br.readLine()));
	}
	
	static String check(String input) {
		Stack<Character> stack = new Stack<>();

			for (int i = 0; i < input.length(); i++) {
				char next = input.charAt(i);
				if (next == '(') {
					stack.push(next);
				} else {
					if (stack.isEmpty()) {
						return "NO";
					} else
						stack.pop();
				}
			}
			if (stack.isEmpty())
				return "YES";
			else
				return "NO";
	}
}